### PR TITLE
#161779863 implement signup and signin feature

### PR DESCRIPTION
**Pull request by Danielshow**
#### What does this PR do?
This PR creates a signup and signin route

#### Description of Task to be completed?
- create signin endpoint
- create login endpoint
- add validations for all fields

#### How should this be manually tested?
```
git clone this branch
npm test

send a request to 
POST api/v1/auth/signup
POST api/vq/auth/signin
```

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#161779863](https://www.pivotaltracker.com/story/show/161779863)

### CODE REVIEW

**Comment by gbols** [comment]
Looks good to me

File: package.json [hunk 1]
```diff
@@ -28,6 +28,8 @@
   "dependencies": {
     "babel-cli": "^6.26.0",
     "babel-preset-env": "^1.7.0",
+    "babel-register": "^6.26.0",
```
Comment (efalayi) [inline package.json:31]: `babel-register` seems to be installed every time 🤔

File: server/controllers/auth/userAuth.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import models from '../../models';
+import password from '../../middlewares/signupValidations/validationHelper';
+
+const { User } = models;
+/**
+ * @description - Performs all auth function
+ */
+class AuthController {
```
Comment (efalayi) [inline server/controllers/auth/userAuth.js:8]: Good job with this 👍

File: server/routes/api/auth/userAuth.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { Router } from 'express';
+import authController from '../../../controllers/auth/userAuth';
+import
+  emailValid from '../../../middlewares/signupValidations/emailValidation';
+import
+  pwdValid from '../../../middlewares/signupValidations/passwordValidation';
```
Comment (efalayi) [inline server/routes/api/auth/userAuth.js]: I think all validations should be in one folder. So we can have `validations/emailValidation`. Why? `emailValidation` can also be required in another action that is not a signup.

File: server/routes/index.js [hunk 4]
```diff
@@ -1,5 +1,9 @@
+import userAuth from './api/auth/userAuth';
+
 const router = require('express').Router();
 
 router.use('/api', require('./api'));
```
Comment (efalayi) [inline server/routes/index.js]: Can this line be removed? I don't think it's required anymore.

File: server/routes/index.js [hunk 5]
```diff
@@ -1,5 +1,9 @@
...
 
+router.use('/api/v1', userAuth);
```
Comment (efalayi) [inline server/routes/index.js:6]: Cool. This is clean 👍

File: tests/users/name.test.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  it('should register a user with a valid username', () => chai.request(url)
+    .post('/api/v1/auth/signup')
+    .send(userObject.users6)
+    .then((res) => {
+      expect(res).to.have.status(200);
+      expect(res.body).to.be.an('Object');
+      expect(res.body.msg).to.be.equals('User created successfully');
+    }));
+
+  it('should gives an error if username field is null', () => chai.request(url)
```
Comment (efalayi) [inline tests/users/name.test.js]: There's a typo here

File: tests/users/name.test.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    .post('/api/v1/auth/signup')
+    .send(userObject.users8)
+    .then((res) => {
+      expect(res).to.have.status(400);
+      expect(res.body).to.be.an('Object');
+      expect(res.body.msg).to.be.equals(
+        'Invalid Username: supply a valid username');
+    }));
+
+  it('should gives error if username is less than 3', () => chai.request(url)
```
Comment (efalayi) [inline tests/users/name.test.js]: There's a typo here

Comment (efalayi) [reply tests/users/name.test.js]: A couple of typos, actually. Please check for typos and correct them.

File: server/middlewares/signupValidations/validationHelper.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  const usernameRegex = /^([a-z0-9!@#$%^&*]){3,20}$/;
+  return usernameRegex.test(name);
+};
+
+/**
+ * @description - This function hashes password
+ * @param {password} password to be hashed
+ * @return {string} hashed password
+ */
+const hashPassword = (password) => {
```
Comment (efalayi) [inline server/middlewares/signupValidations/validationHelper.js]: I'm not sure this is a validation. Should this be here?

File: server/middlewares/signupValidations/passwordValidation.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ */
+class PasswordValidation {
+  /**
+   * @description A middleware for validating email
+   * @param {object} req - get response from body
+   * @param {object} res - response to be sent
+   * @param {object} next - callback function
+   * @returns {object} A response object
+   */
+  static isPasswordCheck(req, res, next) {
```
Comment (efalayi) [inline server/middlewares/signupValidations/passwordValidation.js]: This function name can be better. Your function name should clearly state what it does

**Review by efalayi — COMMENTED** [review]
Great work @Danielshow. Just a couple of changes.
Please attend to them.

**Review by davidshare — APPROVED** [review]
Nice work Daniel, looks good to me.

**Pull request merged**